proposals/0149-migrate-snapshot-epoch-stakes.md

```diff
@@ -0,0 +1,159 @@
+    epoch_authorized_voters: Arc<EpochAuthorizedVoters>,
+}
+
+enum StakesEnum {
+    Accounts(Stakes<StakeAccount>),
+    Delegations(Stakes<Delegation>),
+}
+```
+
+When serializing `EpochStakes` in snapshots, all `StakesEnum` variants first map
+stake entry values to their `Delegation` value before serialization. The goal of
+this proposal is to migrate to a new epoch stakes field which maps stake entry
+values to their full `Stake` value before serialization so that
+`credits_observed` will be be included in the snapshot and available after
+snapshot deserialization.
+
+The proposed `new_epoch_stakes` bank snapshot field will instead serialize a map
+of epochs to `VersionedEpochStakes` structs which can be updated in the future
+to serialize different information if needed. This field will be appended to the
+end of the serialized bank snapshot:
+
+```rust
+struct (BankFieldsToDeserialize | BankFieldsToSerialize) {
+    ..
+    new_epoch_stakes: HashMap<Epoch, VersionedEpochStakes>,
+}
+
+enum VersionedEpochStakes {
+    Current {
+        stakes: Stakes<Stake>,
+        total_stake: u64,
+        node_id_to_vote_accounts: Arc<NodeIdToVoteAccounts>,
+        epoch_authorized_voters: Arc<EpochAuthorizedVoters>,
+    },
+}
+```
+
+### Snapshot Update Rollout
+
+Handling snapshot format changes is always a delicate operation to coordinate
+given that old software releases will not be able to deserialize snapshots from
+new software releases properly. The rollout will require two phases:
+
+1. Introduce support for deserializing the migrated epoch stakes field 
+2. Enable serializing epoch stakes to the new field and phase out the old field
+
+During the first phase, validator software will be updated to attempt to
+deserialize the new epoch stakes field appended at the end of the bank snapshot.
+If the field doesn't exist, validators will continue using the old deserialized
+field. If the new epoch stakes field is deserialized successfully, the epoch
+stakes entries from this field will be merged with entries from the old field.
+
+During the second phase, validator software will be updated to start serializing
+epoch stakes entries to the new epoch stakes field. Note, however, that there
+are 3 different epoch stakes entry variants:
+
+1. Entries created during epoch boundaries which have full stake account data
+2. Entries deserialized from the old snapshot epoch stakes field which only have
+    stake delegation state.
+3. Entries deserialized from the new snapshot epoch stakes field which have full
+    stake state.
+
+Only variants 1 and 3 can be serialized into the new epoch stakes field so any
+variant 2 epoch stakes entries will continue being serialized into the old epoch
+stakes field.
+
+We propose adding a new feature gate `migrate_epoch_stakes_snapshot_field` to
+facilitate the migration by activating phase 2 so that the implementation for
+both phases can be added to the next beta release.
```

Comment:
Alright, you have me convinced that using minor versions is the best way to go, thanks for explaining the context behind your stance here. I've updated the rollout description in the doc, would appreciate another look!

Reply:
Do we consider this a blocker for activating partitioned epoch rewards? I'm concerned about adding an additional 3-4 months before that can be activated. Every epoch boundary is scary.

Reply:
I don't consider it a blocker personally, we can continue keeping the stake program fully locked down to avoid any issues popping up from our usage of stakes cache. Later when this migration is completed, we can allow new stake accounts to be created if we want

Reply:
Okay, cool 👍